plugin/warpctc/warpctc-inl.h

```diff
@@ -0,0 +1,242 @@
+      std::cout << i << "\t" << blob.shape_[i] << std::endl;
+    }
+  }
+  
+  virtual void Forward(const OpContext &ctx,
+		       const std::vector<TBlob> &in_data,
+		       const std::vector<OpReqType> &req,
+		       const std::vector<TBlob> &out_data,
+		       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 2) << "CTCOutput Input: [data, label]";
+    CHECK_EQ(out_data.size(), 1) << "CTCOutput Output: [output]";
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    TBlob data = in_data[warpctc_enum::kData];
+    TBlob out = out_data[warpctc_enum::kOut];
+    Tensor<xpu, 2, DType> data_tensor = data.FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> out_tensor = out.FlatTo2D<xpu, DType>(s);
+    Softmax(out_tensor, data_tensor);
+  }
+
+  virtual void Backward(const OpContext &ctx,
+			const std::vector<TBlob> &out_grad,
+			const std::vector<TBlob> &in_data,
+			const std::vector<TBlob> &out_data,
+			const std::vector<OpReqType> &req,
+			const std::vector<TBlob> &in_grad,
+			const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    TBlob data = out_data[warpctc_enum::kOut];
+    TBlob label = in_data[warpctc_enum::kLabel];
+    CHECK_EQ(data.shape_.ndim(), 2) << "input data shape should be 2 (t*n, p)";
+    ctcComputeInfo info;
+    info.loc = CTC_CPU;
+    info.num_threads = 1;
+    
+    int T = param_.input_length;
+    int minibatch = data.shape_[0] / T;
+    int alphabet_size = data.shape_[1];
+    std::vector<int> input_lengths;
+    for (int i = 0; i < minibatch; i++) {
+      input_lengths.push_back(T);
+    }
+    std::vector<int> label_lengths;
+    for (int i = 0; i < minibatch; i++) {
+      label_lengths.push_back(param_.label_length);
+    }
+
+    size_t cpu_alloc_bytes;
+    throw_on_error(get_workspace_size(label_lengths.data(), input_lengths.data(),
+                                      alphabet_size, input_lengths.size(), info,
+                                      &cpu_alloc_bytes),
+                   "Error: get_workspace_size in inf_test");
+    void* ctc_cpu_workspace = malloc(cpu_alloc_bytes);
+
+    cudaError_t cuda_status;
+    float* activations = static_cast<float*>(data.dptr_);
+    int* flat_labels = static_cast<int*>(label.dptr_);
+    float* cpu_activations = activations;
+    int* cpu_labels = flat_labels;
+
+    if (data.dev_mask_ == gpu::kDevMask) {
+      cpu_activations = (float*) malloc(sizeof(float) * data.Size());
+      cuda_status = cudaMemcpyAsync(cpu_activations, activations, data.Size() * sizeof(float), cudaMemcpyDeviceToHost);
+      CHECK_EQ(cuda_status, cudaSuccess) << "cuda memcpy activation error";
+
+      cpu_labels = (int*) malloc(sizeof(int) * label.Size());
+      cuda_status = cudaMemcpyAsync(cpu_labels, flat_labels, label.Size()*sizeof(int), cudaMemcpyDeviceToHost);
+      CHECK_EQ(cuda_status, cudaSuccess) << "cuda memcpy label error";  
+    }
```

Comment:
If run under GPU mode, I will copy data from GPU to CPU and run computer_ctc_loss.

I fail to run computer_ctc_loss under GPU. It will cause segmentation fault.